grammars/tree-sitter-python.cson

```diff
@@ -66,44 +66,220 @@ scopes:
-  '"await"': 'keyword.control'
-  '"async"': 'keyword.control'
-  '"try"': 'keyword.control'
-  '"except"': 'keyword.control'
-  '"with"': 'keyword.control'
-  '"as"': 'keyword.control'
-  '"finally"': 'keyword.control'
-  '"import"': 'keyword.control'
-  '"from"': 'keyword.control'
-  '"print"': 'keyword.control'
-  '"assert"': 'keyword.control'
+  '"import"': 'keyword.control.import.from.python'
+
+  '"elif"': 'keyword.control.conditional.python'
+  '"else"': 'keyword.control.conditional.python'
+  '"if"': 'keyword.control.conditional.python'
+
+  '"except"': 'keyword.control.exception.python'
+  '"finally"': 'keyword.control.exception.python'
+  '"try"': 'keyword.control.exception.python'
+
+  '"for"': 'keyword.control.repeat.python'
+  '"while"': 'keyword.control.repeat.python'
+
+  '"return"': 'keyword.control.statement.python'
+  '"break"': 'keyword.control.statement.python'
+  '"continue"': 'keyword.control.statement.python'
+  '"raise"': 'keyword.control.statement.python'
+  '"yield"': 'keyword.control.statement.python'
+  '"await"': 'keyword.control.statement.python'
+  '"with"': 'keyword.control.statement.python'
+
+  '"as"': 'keyword.other.python'
+  '"assert"': 'keyword.other.python'
+  '"del"': 'keyword.other.python'
+  '"async"': 'keyword.other.python'
+  '"from"': 'keyword.other.python'
+
+  # Those are handled as "built-in functions" below
+  # '"exec"': 'keyword.deprecated.python'
+  # '"print"': 'keyword.deprecated.python'
 
   '"+"': 'keyword.operator'
   '"-"': 'keyword.operator'
   '"*"': 'keyword.operator'
+  '"**"': 'keyword.operator'
   '"/"': 'keyword.operator'
+  '"//"': 'keyword.operator'
   '"%"': 'keyword.operator'
+  '"@"': 'keyword.operator'
   '"in"': 'keyword.operator.logical.python'
   '"and"': 'keyword.operator.logical.python'
   '"or"': 'keyword.operator.logical.python'
   '"not"': 'keyword.operator.logical.python'
   '"is"': 'keyword.operator.logical.python'
+  '"=="': 'keyword.operator.comparison.python'
+  '"!="': 'keyword.operator.comparison.python'
+  '"<="': 'keyword.operator.comparison.python'
+  '">="': 'keyword.operator.comparison.python'
+  '"<"': 'keyword.operator.comparison.python'
+  '">"': 'keyword.operator.comparison.python'
+  '"="': 'keyword.operator.assignment.python'
+  '"+="': 'keyword.operator.assignment.python'
+  '"-="': 'keyword.operator.assignment.python'
+  '"*="': 'keyword.operator.assignment.python'
+  '"**="': 'keyword.operator.assignment.python'
+  '"/="': 'keyword.operator.assignment.python'
+  '"//="': 'keyword.operator.assignment.python'
+  '"%="': 'keyword.operator.assignment.python'
+  '"@="': 'keyword.operator.assignment.python'
+
+  '"("': 'punctuation.parenthesis'
+  '")"': 'punctuation.parenthesis'
+  '"{"': 'punctuation'
+  '"}"': 'punctuation'
+  '"["': 'punctuation'
+  '"]"': 'punctuation'
+  '"."': 'punctuation'
+  '","': 'punctuation'
+  '";"': 'punctuation'
+  '":"': 'punctuation'
+  '"->"': 'punctuation'
+
+  '"__build_class__"': 'support.function.builtin.python'
```

Comment:
Tree-sitter doesn't output separate tokens for these built-in functions or built-in exceptions; they're just `identifier` in the AST. So we currently don't have a way to highlight these in Tree-sitter mode.

I plan to add an additional layer of regex/string-based highlighting on *top* of the pure AST-based approach, in order to highlight things like escape sequences in strings. Once that system is implemented, it *could* be used for highlighting some of these built-ins. I think it'll require a slightly different syntax, because you'll need to say "When you see an `identifier`, if it matches *this* string, give it *these* scopes". So for now, I think we should take these out.

In general, I'm skeptical of the value of highlighting things like built-in functions, because it will give you a lot of "false positives" when you inevitably have local variables called `id`, `hash`, etc. What are your thoughts on that?

Reply:
In Python it's useful to know whether a name is a built-in precisely because you can unintentionally shadow it with a variable name. All editors I know of show builtins in a different color due to that. This is helpful because some of the builtins aren't used very often *or* might not come to the author's mind in the context of the code they are writing. Seeing that they're assigning to a builtin makes them reconsider their choice of a name.

As for builtin usage, those builtins are semi-keywords, they provide idiomatic APIs (like `len()`, `open()`, `range()`) or types (like `list()`, `dict()`, `set()`, and so on). So in my mind it makes sense to color them differently.

As for lack of support in Tree Sitter for this now, OK, I will take this out, it's rather trivial to re-add in the future. Regex matching would be great for this, then we could also color dunder names (like `__name__`), attribute lookups and calls differently. But that's lower pri and for builtins a straight string match would be perfectly adequate.